pkg/controller/certificatesigningrequests/venafi/venafi.go

```diff
@@ -0,0 +1,206 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package venafi
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/Venafi/vcert/v4/pkg/endpoint"
+	certificatesv1 "k8s.io/api/certificates/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	certificatesclient "k8s.io/client-go/kubernetes/typed/certificates/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/record"
+
+	apiutil "github.com/jetstack/cert-manager/pkg/api/util"
+	cmapi "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
+	experimentalapi "github.com/jetstack/cert-manager/pkg/apis/experimental/v1alpha1"
+	controllerpkg "github.com/jetstack/cert-manager/pkg/controller"
+	"github.com/jetstack/cert-manager/pkg/controller/certificatesigningrequests"
+	"github.com/jetstack/cert-manager/pkg/controller/certificatesigningrequests/util"
+	venaficlient "github.com/jetstack/cert-manager/pkg/issuer/venafi/client"
+	venafiapi "github.com/jetstack/cert-manager/pkg/issuer/venafi/client/api"
+	logf "github.com/jetstack/cert-manager/pkg/logs"
+	"github.com/jetstack/cert-manager/pkg/util/pki"
+	utilpki "github.com/jetstack/cert-manager/pkg/util/pki"
+)
+
+const (
+	CSRControllerName = "certificatesigningrequests-issuer-venafi"
+)
+
+// Venafi is a Kubernetes CertificateSigningRequest controller, responsible for
+// signing CertificateSigningRequests that reference a cert-manager Venafi
+// Issuer or ClusterIssuer
+type Venafi struct {
+	issuerOptions controllerpkg.IssuerOptions
+	secretsLister corelisters.SecretLister
+	certClient    certificatesclient.CertificateSigningRequestInterface
+	recorder      record.EventRecorder
+
+	clientBuilder venaficlient.VenafiClientBuilder
+}
+
+func init() {
+	controllerpkg.Register(CSRControllerName, func(ctx *controllerpkg.Context) (controllerpkg.Interface, error) {
+		return controllerpkg.NewBuilder(ctx, CSRControllerName).
+			For(certificatesigningrequests.New(apiutil.IssuerVenafi, NewVenafi(ctx))).
+			Complete()
+	})
+}
+
+func NewVenafi(ctx *controllerpkg.Context) *Venafi {
+	return &Venafi{
+		issuerOptions: ctx.IssuerOptions,
+		secretsLister: ctx.KubeSharedInformerFactory.Core().V1().Secrets().Lister(),
+		certClient:    ctx.Client.CertificatesV1().CertificateSigningRequests(),
+		recorder:      ctx.Recorder,
+		clientBuilder: venaficlient.New,
+	}
+}
+
+// Sign attempts to sign the given CertificateSigningRequest based on the
+// provided Venafi Issuer or ClusterIssuer. This function will update the resource
+// if signing was successful. Returns an error which, if not nil, should
+// trigger a retry.
+// Since this signer takes some time to sign the request, this controller will
+// set a "pick ID" annotation value that is used to fetch the latest state of
+// the request in subsequent re-syncs.
```

Comment:
question:

> used to fetch the latest state of the request in subsequent re-syncs.

Do we resync periodically e.g. every 10 seconds to check if the certificate has been signed using the `picking_id`? Or do we rely on the default 10 hours resync period for that?

Reply:
We currently rely on returning an error, and letting the rate limiting queue take its course. 

https://github.com/jetstack/cert-manager/pull/4108/files/2e44f5d42efbf86fac4c87d6544b0894c95b0a1b#diff-3b5f26d65b2aa51ae2ada8f64c701cb765ff4bc7494740e36353f55cf5841cf8R174

This is the same behavior as CertificateRequest https://github.com/jetstack/cert-manager/blob/88e85d07253f8d8a99628a2220c7f751a23ef4f1/pkg/controller/certificaterequests/venafi/venafi.go#L152

Reply:
suggestion: explain how the resyncs are scheduled, eg

```go
// Since this signer takes some time to sign the request, this controller will
// set a "pick ID" annotation value that is used to fetch the latest state of
// the request in subsequent re-syncs. The re-syncs are triggered by using the
// workqueue's back-off mechanism.
```

Reply:
Will add a comment